File: smart-chits-app Backend/app/services/chit_service.py
from sqlalchemy.orm import Session
from app.models.chit_model import Chit
from app.models.member_model import Member
from app.services.notification_service import create_notification


# ✅ Create chit
def create_chit(db: Session, data):
    chit = Chit(**data.model_dump())
    db.add(chit)
    db.commit()
    db.refresh(chit)

    create_notification(db, None, f"Payment due for chit '{chit.name}' month {chit.current_month}")
    return chit


# ✅ Get all chits
def get_all_chits(db: Session):
    return db.query(Chit).all()


# ✅ Join chit
def join_chit(db: Session, user_id: int, chit_id: int):

    chit = db.query(Chit).filter(Chit.id == chit_id).first()
    if not chit:
        raise Exception("Chit not found")

    # check already joined
    existing = db.query(Member).filter(
        Member.user_id == user_id,
        Member.chit_id == chit_id
    ).first()

    if existing:
        raise Exception("Already joined")

    # check limit
    count = db.query(Member).filter(Member.chit_id == chit_id).count()
    if count >= chit.total_members:
        raise Exception("Chit is full")

    member = Member(user_id=user_id, chit_id=chit_id)
    db.add(member)
    db.commit()
    db.refresh(member)

    return member

File: smart-chits-app UI/smart-chits-frontend/src/services/chitService.js
import { fetcher } from './api'

// Get all chits
export async function getAllChits() {
  const response = await fetcher('/chits/')
  return response
}

// Create a new chit (admin only)
export async function createChit(chitData) {
  const response = await fetcher('/chits/create', {
    method: 'POST',
    body: JSON.stringify(chitData),
  })
  return response
}

// Add member to chit (admin only)
export async function addMemberToChit(userId, chitId) {
  const response = await fetcher('/chits/add-member', {
    method: 'POST',
    body: JSON.stringify({ user_id: userId, chit_id: chitId }),
  })
  return response
}

// Get chit details
export async function getChitDetails(chitId) {
  const response = await fetcher(`/chits/${chitId}`)
  return response
}

// Join a chit (normal user)
export async function joinChit(chitId) {
  const response = await fetcher(`/chits/${chitId}/join`, {
    method: 'POST',
  })
  return response
}

// Check if user is already a member of a chit
export async function checkMembership(chitId) {
  const response = await fetcher(`/chits/${chitId}/is-member`)
  return response
}

// Get user's joined chits
export async function getMyChits() {
  const response = await fetcher('/admin/my-chits')
  return response
}
